### PR TITLE
Add boundary condition support to Creators::Interval

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Adds support for specifying boundary conditions at runtime and adding them to the domain to the interval domain creator.

Some notes:
- In order to not break develop and not have one 15,000 line PR (not an exaggeration...) I have to support both having and not having boundary conditions specified in the systems. Once all the systems (both elliptic and evolution) are migrated over to supporting boundary conditions, we can clean up the code. This is noted in #2779 along with some related cleanups
- There are some "extra" or "unexpected" changes, such as the need to register boundary conditions in some of the elliptic test executables.
- Once folks have had a chance to look at the interface, I'll open up PRs for all the remaining domain creators. I have not added boundary condition support for all conceivable cases of option values to all domain creators because I'm not sure we will ever use some of them, and if we do, adding them isn't so difficult that whoever wants that configuration couldn't add the support.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Domain/Creators/Interval.hpp [hunk 1]
```diff
@@ -69,19 +71,56 @@ class Interval : public DomainCreator<1> {
...
+      tmpl::push_back<
+          basic_options,
+          LowerBoundaryCondition<
+              domain::BoundaryConditions::get_boundary_conditions_base<
+                  typename Metavariables::system>>,
+          UpperBoundaryCondition<
+              domain::BoundaryConditions::get_boundary_conditions_base<
+                  typename Metavariables::system>>>,
+      basic_options>;
+
```
Comment (nilsvu) [inline src/Domain/Creators/Interval.hpp]: How about this pattern:

```cpp
template <typename Metavariables>
using options = tmpl::append<
    tmpl::list<LowerBound, UpperBound, IsPeriodicIn, InitialRefinement,
               InitialGridPoints, TimeDependence>,
    tmpl::conditional_t<
        domain::BoundaryConditions::has_boundary_conditions_base_v<
            typename Metavariables::system>,
        tmpl::list<LowerBoundaryCondition<
                       domain::BoundaryConditions::get_boundary_conditions_base<
                           typename Metavariables::system>>,
                   UpperBoundaryCondition<
                       domain::BoundaryConditions::get_boundary_conditions_base<
                           typename Metavariables::system>>>,
        tmpl::list<>>>;
```

I find this easier to understand, and it's easier to extend the list with more conditionals.

File: tests/Unit/Domain/Creators/Test_AlignedLattice.cpp [hunk 2]
```diff
@@ -44,16 +46,18 @@ void test_aligned_blocks(
...
+      TestHelpers::domain::BoundaryConditions::
+          MetavariablesWithBoundaryConditions<1>>(
+      "AlignedLattice:\n"
+      "  BlockBounds: [[0.1, 2.6, 5.1, 5.2, 7.2]]\n"
+      "  IsPeriodicIn: [false]\n"
+      "  InitialGridPoints: [3]\n"
+      "  InitialLevels: [2]\n"
+      "  RefinedLevels: []\n"
+      "  RefinedGridPoints: []\n"
+      "  BlocksToExclude: []\n");
```
Comment (nilsvu) [inline tests/Unit/Domain/Creators/Test_AlignedLattice.cpp:60]: This looks like it belongs in the next PR

File: tests/Unit/Domain/Creators/CMakeLists.txt [hunk 3]
```diff
@@ -26,5 +26,15 @@ add_test_library(
...
   "${LIBRARY_SOURCES}"
-  "DomainCreators;Domain;DomainHelpers;Test_Domain"
+  ""
+  )
+
+target_link_libraries(
+  ${LIBRARY}
+  PRIVATE
+  Domain
+  DomainBoundaryConditionsHelpers
```
Comment (nilsvu) [inline tests/Unit/Domain/Creators/CMakeLists.txt:37]: Add `DomainBoundaryConditions`

File: tests/Unit/Domain/Creators/Test_Interval.cpp [hunk 4]
```diff
@@ -83,80 +90,174 @@ void test_interval() {
...
+    test_interval_construction(
+        periodic_interval, lower_bound, upper_bound, grid_points,
+        refinement_level,
+        std::vector<DirectionMap<1, BlockNeighbor<1>>>{
+            {{Direction<1>::lower_xi(), {0, aligned_orientation}},
+             {Direction<1>::upper_xi(), {0, aligned_orientation}}}},
+        std::vector<std::unordered_set<Direction<1>>>{{}}, std::tuple<>{}, {},
+        {});
+  }
+  { INFO("With boundary condition"); }
```
Comment (nilsvu) [inline tests/Unit/Domain/Creators/Test_Interval.cpp]: This looks like it wants something to do :)

File: tests/Unit/Domain/Creators/Test_Interval.cpp [hunk 5]
```diff
@@ -83,80 +90,174 @@ void test_interval() {
...
+      Direction<1>::upper_xi(), 0);
+  const std::vector<std::unordered_set<Direction<1>>>
+      expected_external_boundaries{
+          {Direction<1>::lower_xi(), Direction<1>::upper_xi()}};
+
+  // for periodic domains:
+  const std::vector<DirectionMap<1, BlockNeighbor<1>>> expected_neighbors{
+      {{Direction<1>::lower_xi(), {0, {}}},
+       {Direction<1>::upper_xi(), {0, {}}}}};
+  {{INFO("Interval factory time independent, no boundary condition");
```
Comment (nilsvu) [inline tests/Unit/Domain/Creators/Test_Interval.cpp]: Please clang-format this file

Also I think you can remove the outer scopes

File: tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.yaml [hunk 6]
```diff
@@ -20,6 +20,15 @@ DomainCreator:
...
     TimeDependence: None
+    BoundaryConditions:
+      LowerBoundary:
+        TestBoundaryCondition:
+          Direction: lower-xi
+          BlockId: 0
+      UpperBoundary:
+        TestBoundaryCondition:
+          Direction: upper-xi
+          BlockId: 0
```
Comment (nilsvu) [inline tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.yaml]: I'm fine with making these changes, though you could get around that by specifying the test-system without boundary conditions. We may end up with a way of avoiding dummy boundary conditions for things like this or ExportCoordinates or ApplyObservers, so it might be easier to do what you did for ExportCoordinates and just add the empty system here. Up to you.

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp [hunk 7]
```diff
@@ -109,15 +111,17 @@ void test_rotated_intervals_factory() {
...
+      DomainCreator<1>, domain::OptionTags::DomainCreator<1>,
+      TestHelpers::domain::BoundaryConditions::
+          MetavariablesWithoutBoundaryConditions<1>>(
+      "RotatedIntervals:\n"
+      "  LowerBound: [0.0]\n"
+      "  Midpoint:   [0.5]\n"
+      "  UpperBound: [1.0]\n"
+      "  IsPeriodicIn: [True]\n"
+      "  InitialGridPoints: [[3,2]]\n"
+      "  InitialRefinement: [2]\n");
```
Comment (nilsvu) [inline tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:124]: This also looks like it belongs in the next PR

**Review by nilsvu — CHANGES_REQUESTED** [review]
Please squash these directly

File: src/Domain/Creators/Interval.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Domain/Creators/Interval.hpp]: I find this much harder to understand. Since this is getting removed after all the systems have boundary conditions anyway, I don't see it as being worth the effort to change in every domain creator.

If we want to support domains without systems we should probably completely rethink afterwards, since this approach is admittedly tedious but I couldn't think of a better way.

File: tests/Unit/Domain/Creators/Test_AlignedLattice.cpp (hunk 2 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/Creators/Test_AlignedLattice.cpp:60]: No, factory creation needs to be able to create all domain creators, and `Interval` no longer supports `void` as a metavariable

File: tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.yaml (hunk 6 above)
Comment (nilsdeppe) [reply tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.yaml]: I made the change anticipating you'd want to add boundary conditions into these tests afterward :)

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp (hunk 7 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:124]: No, factory creation needs to be able to create all domain creators, and `Interval` no longer supports `void` as a metavariable

**Comment by nilsdeppe** [comment]
Rebased and squashed. Thanks for reviewing :)

File: tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.yaml (hunk 6 above)
Comment (nilsvu) [reply tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.yaml]: Ah I see. Would you mind going with the empty-system approach for now then? The linear solver tests won't use boundary conditions because they work with hard-coded matrices.

Comment (nilsdeppe) [reply tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.yaml]: sure!

**Comment by nilsdeppe** [comment]
Rebased and removed the boundary conditions from the elliptic tests. Thanks for the review :)

**Review by nilsvu — DISMISSED** [review]

File: src/Domain/Creators/Interval.cpp [hunk 8]
```diff
@@ -49,12 +52,36 @@ Interval::Interval(
 }
 
 Domain<1> Interval::create_domain() const noexcept {
+  std::vector<DirectionMap<
+      1, std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>>
+      boundary_conditions_all_blocks{};
+  if (lower_boundary_condition_ != nullptr) {
+    if (upper_boundary_condition_ == nullptr)  {
+      ERROR(
```
Comment (fmahebert) [inline src/Domain/Creators/Interval.cpp]: This ERROR will trigger when only a lower boundary condition is specified. I suggest adding a matching ERROR for the opposite case, i.e. when only the upper boundary condition is specified.

File: src/Domain/Creators/Interval.cpp [hunk 9]
```diff
@@ -49,12 +52,36 @@ Interval::Interval(
...
+          "Either both the lower and upper boundary need to be specified or "
+          "neither.");
+    }
+    if (is_periodic_in_x_[0]) {
+      ERROR(
```
Comment (fmahebert) [inline src/Domain/Creators/Interval.cpp]: Likewise, should check periodic isn't paired with only an upper boundary condition

File: src/Executables/ExportCoordinates/ExportCoordinates.hpp [hunk 10]
```diff
@@ -180,6 +180,9 @@ template <size_t Dim>
 struct Metavariables {
   static constexpr size_t volume_dim = Dim;
   static constexpr bool local_time_stepping = false;
+  // all the domain creators need a system now. Currently don't require boundary
+  // conditions, so can wait to add placeholders for those.
```
Comment (fmahebert) [inline src/Executables/ExportCoordinates/ExportCoordinates.hpp]: [minor] I slightly dislike "now", "currently", because it's not clear to a future reader what times are being referred to. I suggest simply saying: "A placeholder system for the domain creator."

File: src/Domain/Creators/Interval.cpp [hunk 11]
```diff
@@ -49,12 +52,36 @@ Interval::Interval(
 }
 
 Domain<1> Interval::create_domain() const noexcept {
+  std::vector<DirectionMap<
+      1, std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>>
+      boundary_conditions_all_blocks{};
+  if (lower_boundary_condition_ != nullptr) {
+    if (upper_boundary_condition_ == nullptr)  {
```
Comment (fmahebert) [inline src/Domain/Creators/Interval.cpp]: Question to check my understanding of the interface: what does the user do in the input file to get a nullptr here? Is it by leaving the option blank, like this?
```
BoundaryConditions:
  LowerBoundary:
  UpperBoundary:
```

**Review by fmahebert — COMMENTED** [review]
Some thoughts from a quick pass

File: src/Domain/Creators/Interval.cpp (hunk 11 above)
Comment (nilsdeppe) [reply src/Domain/Creators/Interval.cpp]: This is an annoying one that I haven't settled yet... I think the best option is to remove the `IsPeriodicIn` stuff completely and add a system-agnostic `Periodic` condition that is checked.

File: tests/Unit/Domain/Creators/Test_Interval.cpp [hunk 12]
```diff
@@ -83,80 +90,198 @@ void test_interval() {
...
+        std::make_tuple(
+            std::pair<std::string,
+                      domain::FunctionsOfTime::PiecewisePolynomial<2>>{
+                "TranslationX",
+                {1.0, std::array<DataVector, 3>{{{0.0}, {2.3}, {0.0}}}, 10.0}}),
+        make_vector_coordinate_map_base<Frame::Grid, Frame::Inertial>(
+            CoordinateMaps::TimeDependent::Translation{"TranslationX"}),
+        expected_boundary_conditions);
   }
 }
```
Comment (kidder) [inline tests/Unit/Domain/Creators/Test_Interval.cpp:324]: test the new ERRORs you introduced

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
I have pushed commits (won't pass CI until after Blaze 3.8 is merged) that do the following:
- a new commit that adds a `Periodic` base class into the `Domain` that any system-specific periodic boundary conditions must inherit from to mark them as periodic. I intentionally did not do a virtual function because somebody will inevitably copy-pasta and create a "periodic" condition that isn't actually periodic. The design in this PR makes that rather difficult, and the tests will catch it.
- added a fixup commit that pretty much completely reworks `Interval` to support periodic boundary conditions with the new boundary condition interface. I've tested this in evolutions with scalar wave, and it behaves as expected. I'll propagate to the other domain creators once the design is settled, but I'm happy with this.

File: src/Domain/Creators/Interval.hpp [hunk 13]
```diff
@@ -76,51 +76,61 @@ class Interval : public DomainCreator<1> {
...
-    using group = BoundaryConditions;
   };
   template <typename BoundaryConditionsBase>
   struct LowerBoundaryCondition {
-    static std::string name() noexcept { return "LowerBoundary"; }
+    static std::string name() noexcept { return "LowerBoundaryCondition"; }
     static constexpr Options::String help =
         "Options for the boundary condition applied at the lower boundary.";
     using type = std::unique_ptr<BoundaryConditionsBase>;
-    using group = BoundaryConditions;
```
Comment (nilsvu) [inline src/Domain/Creators/Interval.hpp]: Why did you remove this group? I liked the `BoundaryConditions:` section in the input file.

File: src/Domain/Creators/Interval.hpp [hunk 14]
```diff
@@ -76,51 +76,61 @@ class Interval : public DomainCreator<1> {
...
-      basic_options>;
+  using options = tmpl::append<
+      common_options,
+      tmpl::conditional_t<
+          domain::BoundaryConditions::has_boundary_conditions_base_v<
+              typename Metavariables::system>,
+          options_boundary_conditions<typename Metavariables::system>,
+          options_periodic>,
+      // Need TimeDependence at the end to not break the already broken python
+      // bindings. This is so that the constructors can default them to nullptr.
```
Comment (nilsvu) [inline src/Domain/Creators/Interval.hpp]: `git apply` this, then you can pass `nullptr`s to the constructor in the order you want:

```diff
diff --git a/src/Domain/Creators/Python/Interval.cpp b/src/Domain/Creators/Python/Interval.cpp
index e3027692f..0c7a80af4 100644
--- a/src/Domain/Creators/Python/Interval.cpp
+++ b/src/Domain/Creators/Python/Interval.cpp
@@ -17,9 +17,18 @@ namespace creators {
 namespace py_bindings {
 void bind_interval(py::module& m) {  // NOLINT
   py::class_<Interval, DomainCreator<1>>(m, "Interval")
-      .def(py::init<std::array<double, 1>, std::array<double, 1>,
-                    std::array<bool, 1>, std::array<size_t, 1>,
-                    std::array<size_t, 1>>(),
+      .def(py::init(
+               [](std::array<double, 1> lower_x, std::array<double, 1> upper_x,
+                  std::array<bool, 1> is_periodic_in_x,
+                  std::array<size_t, 1> initial_refinement_level_x,
+                  std::array<size_t, 1> initial_number_of_grid_points_in_x) {
+                 return Interval{std::move(lower_x),
+                                 std::move(upper_x),
+                                 std::move(is_periodic_in_x),
+                                 std::move(initial_refinement_level_x),
+                                 std::move(initial_number_of_grid_points_in_x),
+                                 nullptr};
+               }),
            py::arg("lower_x"), py::arg("upper_x"), py::arg("is_periodic_in_x"),
            py::arg("initial_refinement_level_x"),
            py::arg("initial_number_of_grid_points_in_x"));
```

File: src/Domain/Creators/Interval.hpp (hunk 13 above)
Comment (nilsdeppe) [reply src/Domain/Creators/Interval.hpp]: I had switched to using `Alternatives` at an intermediate stage. I agree, it's nice to have the group. I'll restore that. Thanks!

**Comment by nilsdeppe** [comment]
- rebased on develop
- added new fixup that restores the group, and removed the ` = nullptr` argument (though I may need to undo this if we want to support compile-time no time dependence)

File: src/Domain/BoundaryConditions/Periodic.hpp [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace domain::BoundaryConditions {
+/// Mark a boundary condition as being periodic.
+///
+/// Periodic boundary conditions shouldn't require any implementation outside of
+/// a check in the domain creator using the `is_periodic()` function to
+/// determine what boundaries are periodic.
```
Comment (fmahebert) [inline src/Domain/BoundaryConditions/Periodic.hpp]: Consider appending a sentence like "Across each matching pair of periodic boundary conditions, the domain creator should specify that the DG elements are neighbors of each other."

My thinking is to make sure users understand the domain creator has to do something with the information... you can't just put a periodic BC in the input file and expect magic.

File: src/Domain/Creators/Interval.hpp [hunk 16]
```diff
@@ -75,18 +75,20 @@ class Interval : public DomainCreator<1> {
     static constexpr Options::String help = "The boundary conditions to apply.";
   };
   template <typename BoundaryConditionsBase>
-  struct UpperBoundaryCondition {
-    static std::string name() noexcept { return "UpperBoundaryCondition"; }
+  struct UpperBoundary {
+    static std::string name() noexcept { return "UpperBoundary"; }
```
Comment (fmahebert) [inline src/Domain/Creators/Interval.hpp:79]: Is this `name` string needed, if the name matches the struct name? (also for Lower case below)

File: src/Domain/Creators/Interval.cpp [hunk 17]
```diff
@@ -27,34 +28,96 @@ struct Logical;
...
+      time_dependence_(std::move(time_dependence)) {
+  if (time_dependence_ == nullptr) {
+    time_dependence_ =
+        std::make_unique<domain::creators::time_dependence::None<1>>();
+  }
+  using domain::BoundaryConditions::is_periodic;
+  if ((is_periodic(lower_boundary_condition_) and
+       not is_periodic(upper_boundary_condition_)) or
+      (not is_periodic(lower_boundary_condition_) and
+       is_periodic(upper_boundary_condition_))) {
```
Comment (fmahebert) [inline src/Domain/Creators/Interval.cpp]: I think the shorter expression `if (is_periodic(lower) != is_periodic(upper))` has the same result? Certainly shorter to take in, though debatable if it's clearer or not... up to you if you like this or not :D

File: tests/Unit/Domain/Creators/Test_Interval.cpp [hunk 18]
```diff
@@ -83,80 +90,236 @@ void test_interval() {
...
-      refinement_level,
-      std::vector<DirectionMap<1, BlockNeighbor<1>>>{
-          {{Direction<1>::lower_xi(), {0, aligned_orientation}},
-           {Direction<1>::upper_xi(), {0, aligned_orientation}}}},
-      std::vector<std::unordered_set<Direction<1>>>{{}}, std::tuple<>{}, {});
+    const std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>
+        periodic = std::make_unique<TestHelpers::domain::BoundaryConditions::
+                                        TestPeriodicBoundaryCondition<1>>();
+    const creators::Interval periodic_interval{
+        lower_bound,    upper_bound,           refinement_level[0],
```
Comment (fmahebert) [inline tests/Unit/Domain/Creators/Test_Interval.cpp:155]: Formatting troubles here and a few lines above

**Comment by fmahebert** [comment]
Oops, meant to also say: fixups look good to me, you have my okay to squash them as well as any changes to my immediately above batch of comments.

File: src/Domain/Creators/Interval.hpp (hunk 16 above)
Comment (nilsvu) [reply src/Domain/Creators/Interval.hpp:79]: I liked the original setup here best, where the tag was named `UpperBoundaryCondition` and its name was overridden to `UpperBoundary`. I suggest to go with that when squashing this. Up to you.

File: src/Domain/Creators/Python/Interval.cpp [hunk 19]
```diff
@@ -15,9 +15,19 @@ namespace py = pybind11;
 namespace domain::creators::py_bindings {
 void bind_interval(py::module& m) {  // NOLINT
   py::class_<Interval, DomainCreator<1>>(m, "Interval")
-      .def(py::init<std::array<double, 1>, std::array<double, 1>,
-                    std::array<size_t, 1>, std::array<size_t, 1>,
-                    std::array<bool, 1>>(),
+      .def(py::init(
+               [](std::array<double, 1> lower_x, std::array<double, 1> upper_x,
+                  std::array<bool, 1> is_periodic_in_x,
```
Comment (nilsvu) [inline src/Domain/Creators/Python/Interval.cpp]: please move the `is_periodic_in_x` arg to the end, or move the `py::arg("is_periodic_in_x")` below to the correct position.

File: src/Domain/Creators/Python/Interval.cpp [hunk 20]
```diff
@@ -15,9 +15,19 @@ namespace py = pybind11;
...
+                  std::array<size_t, 1> initial_refinement_level_x,
+                  std::array<size_t, 1> initial_number_of_grid_points_in_x) {
+                 return domain::creators::Interval{
+                     std::move(lower_x),
+                     std::move(upper_x),
+                     std::move(initial_refinement_level_x),
+                     std::move(initial_number_of_grid_points_in_x),
+                     std::move(is_periodic_in_x),
```
Comment (nilsvu) [inline src/Domain/Creators/Python/Interval.cpp]: clang-tidy want's you to remove the trivial `moves here

**Review by nilsvu — CHANGES_REQUESTED** [review]
LGTM, please squash, including these

File: src/Domain/Creators/Interval.cpp (hunk 17 above)
Comment (kidder) [reply src/Domain/Creators/Interval.cpp]: I like the shorter expression

**Review by kidder — COMMENTED** [review]
I have no further requested changes

File: tests/Unit/Domain/Creators/Test_Interval.cpp (hunk 18 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/Creators/Test_Interval.cpp:155]: If you are referring to the large spaces, this is clang-format lining up the commas. If you mean something else, I'm not sure what since running clang-format over the entire file doesn't change anything

**Comment by nilsdeppe** [comment]
Okay, squashed and rebased. Thanks for the reviews!

File: tests/Unit/Domain/Creators/Test_Interval.cpp (hunk 18 above)
Comment (fmahebert) [reply tests/Unit/Domain/Creators/Test_Interval.cpp:155]: If this is the output of clang-format, then so be it... though I've never so strongly disagreed with it

**Review by fmahebert — APPROVED** [review]

File: tests/Unit/Domain/Creators/Test_Interval.cpp (hunk 18 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/Creators/Test_Interval.cpp:155]: And I have yet to figure out when it decides this is the "right" formatting vs. just making a long list...

**Review by kidder — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]
